client/service/src/builder.rs

```diff
@@ -874,22 +874,35 @@ ServiceBuilder<
 			let network_state_info: Arc<dyn NetworkStateInfo + Send + Sync> = network.clone();
 			let is_validator = config.roles.is_authority();
 
-			let events = client.import_notification_stream()
-				.for_each(move |notification| {
+			let (import_stream, finality_stream) = (
+				client.import_notification_stream().map(|n| ChainEvent::Canonical {
+					id: BlockId::Hash(n.hash),
+					header: n.header,
+					retracted: n.retracted,
+				}),
+				client.finality_notification_stream().map(|n| ChainEvent::Finalized {
+					hash: n.hash
+				})
+			);
+			let events = futures::stream::select(import_stream, finality_stream)
+				.for_each(move |event| {
 					let txpool = txpool.upgrade();
 
 					if let Some(txpool) = txpool.as_ref() {
-						let future = txpool.maintain(
-							&BlockId::hash(notification.hash),
-							&notification.retracted,
-						);
+						let future = txpool.maintain(&event);
```

Comment:
@NikVolf can't pass by value because the event will be reused by the off-chain worker.

Reply:
If yo want, you can do (since you reuse header only)
```rust
enum ChainEvent<'a> { 
  Canonical: { 
     header: &'a Header, 
     ... 
  },
  Finalized { ... }
}
```

Reply:
So previously the trait took references to its arguments, I think there's nothing wrong with leaving it as it is. If other systems want to consume the `ChainEvent` stream they would also take references instead of by value which would lead to unnecessary cloning.